src/lib/journey/stages/qr-code.svelte

```diff
@@ -0,0 +1,198 @@
+  import Alert from '$components/primitives/alert/alert.svelte';
+  import Button from '$components/primitives/button/button.svelte';
+  import Form from '$components/primitives/form/form.svelte';
+  import ClipboardIcon from '$components/icons/clipboard-icon.svelte';
+  import MobileIcon from '$components/icons/mobile-icon.svelte';
+  import Link from '$components/primitives/link/link.svelte';
+  import { convertStringToKey } from '$journey/stages/_utilities/step.utilities';
+  import { styleStore as style } from '$lib/style.store';
+
+  // Types
+  import type {
+    CallbackMetadata,
+    StageFormObject,
+    StageJourneyObject,
+    StepMetadata,
+  } from '$journey/journey.interfaces';
+  import type { Maybe } from '$lib/interfaces';
+  import CallbackMapper from '$journey/_utilities/callback-mapper.svelte';
+
+  export let componentStyle: 'app' | 'inline' | 'modal';
+  export let form: StageFormObject;
+  export let formEl: HTMLFormElement | null = null;
+  export let journey: StageJourneyObject;
+  export let metadata: Maybe<{
+    callbacks: CallbackMetadata[];
+    step: StepMetadata;
+  }>;
+  export let step: FRStep;
+
+  let alertNeedsFocus = false;
+  let buttons: { value: string; text: string }[];
+  let formMessageKey = '';
+  let modifiedCallbacks: FRCallback[] = [];
+  let qrCodeCanvas: HTMLCanvasElement;
+  let qrCodeUrl: string = '';
+  let qrCodeEl: HTMLInputElement;
+
+  function determineSubmission() {
+    // TODO: the below is more strict; all self-submitting cbs have to complete before submitting
+    // if (stepMetadata.isStepSelfSubmittable && isStepReadyToSubmit(callbackMetadataArray)) {
+
+    // The below variation is more liberal first self-submittable cb to call this wins.
+    if (metadata?.step?.derived.isStepSelfSubmittable()) {
+      form?.submit();
+    }
+  }
+
+  async function copyUrl() {
+    qrCodeEl.select();
+    document.execCommand('copy');
+  }
+
+  afterUpdate(() => {
+    alertNeedsFocus = !!form?.message;
+  });
+
+  onMount(() => {
+    async function renderQrCodeImage() {
+      try {
+        // Dynamically import QR Code module to reduce initial load when not needed
+        const qrCodeModule = await import('qrcode');
+        qrCodeUrl = FRQRCode.getQRCodeData(step).uri;
+
+        qrCodeModule.toCanvas(
+          qrCodeCanvas,
+          qrCodeUrl,
+          // Properties required for ForgeRock QR Codes
+          { errorCorrectionLevel: 'L', version: 20, width: 400 },
+          function (error) {
+            if (error) {
+              form.message = interpolate('qrCodeFailedToRender');
+              console.error(error);
+            }
+          },
+        );
+      } catch (err) {
+        form.message = interpolate('qrCodeImportFailure');
+        console.error('Failed to import QR Code library');
+      }
+    }
+
+    renderQrCodeImage();
+  });
+
+  $: {
+    // Assign stage value to `QRCode`
+    if (metadata) {
+      metadata.step.derived.stageName = 'QRCode';
+    }
```

Comment:
I'm planning on removing this in favor of pulling out the `PollingWaitCallback` from the array and rendering it in isolation so I can pass in a modifier.